eip-4626.md

```diff
@@ -0,0 +1,148 @@
+author: Joey Santoro (@joeysantoro), t11s (@transmissions11), Jet Jadeja (@JetJadeja)
+discussions-to: https://ethereum-magicians.org/t/eip-4626-yield-bearing-vault-standard/7900
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-12-22
+---
+
+## Abstract
+
+The following standard allows for the implementation of a standard API for tokenized vaults with a single underlying ERC20 within smart contracts. This standard is an extension on the ERC20 token that provides basic functionality for depositing and withdrawing tokens and reading balances.
+
+## Motivation
+
+Tokenized vaults have a lack of standardization leading to diverse implementation details. Some various examples include lending markets (Compound, Aave, Fuse), aggregators (Yearn, Rari Vaults, Idle), and intrinsically interest bearing tokens (xSushi). This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards. This forces each protocol to implement their own adapters which are error prone and waste development resources.
+
+A standard for tokenized vaults will allow for a similar cambrian explosion to ERC-20, unlocking access to yield and other strategies in a variety of applications with little specialized effort from developers.
+
+
+## Specification
+
+All tokenized vaults MUST implement ERC20. If a vault is to be non-transferrable, it MAY revert on calls to transfer or transferFrom. The ERC20 operations balanceOf, transfer, totalSupply, etc. operate on the vault "shares" which represent ownership in the underlying.
+
+### Methods
+
+#### deposit
+
+`function deposit(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Deposits `_value` tokens into the vault and grants ownership of them to `_to`.
+
+MUST emit the `Deposit` event.
+
+#### withdraw
+
+`function withdraw(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Withdraws `_value` tokens from the vault and transfers them to `_to`. 
+
+MUST emit the `Withdraw` event.
+
+#### withdrawFrom
+
+`function withdrawFrom(address _from, address _to, uint256 _value) public returns (uint256 _shares)`
+
+Withdraws `_value` tokens from owner `_from` and transfers them to `_to`. 
+
+MUST share the same authorization mechanism as transferFrom.
+
+MUST emit the `Withdraw` event.
+
+#### redeem
+
+`function redeem(address _to, uint256 _shares) public returns (uint256 _value)`
+
+Redeems a specific number of `_shares` for underlying tokens and transfers them to `_to`. 
+
+MUST emit the `Withdraw` event.
+
+#### redeemFrom
+
+`function redeemFrom(address _from, address _to, uint256 _shares) public returns (uint256 _value)`
+
+Redeems a specific number of `_shares` from owner `_from` for underlying tokens and transfers them to `_to`. 
+
+MUST share the same authorization mechanism as transferFrom.
+
+MUST emit the `Withdraw` event.
+
+#### totalHoldings
+`function totalHoldings() public view returns (uint256)`
```

Comment:
Can we make this `totalUnderlying()` to conform with `balanceOfUnderlying()`, `calculateUnderlying()`, `underlying()` etc

Reply:
Yes we will change this